pkg/apis/v1alpha1/vpcnetworkconfiguration_types.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// SubnetType describes the subnet type.
+type SubnetType string
+
+const (
+	SubnetTypePublic  SubnetType = "public"
+	SubnetTypePrivate SubnetType = "private"
+)
+
+// VPCNetworkConfigurationSpec defines the desired state of VPCNetworkConfiguration
+type VPCNetworkConfigurationSpec struct {
```

Comment:
> Checked with Kevin that the VPC API can have optional values, but they do not have default values or inherit settings of a higher level object from the VMODL definition. But in the implementation of the VAPI, we can set default values for optional fields if they are not set.

I know VMODL does not support default value definition. I meant the API implementation. Does the VAPI implementation usually sets optional fields with inherited values from a higher level object? PMs might not give inputs on API. Let us first check any convention in VAPI.

Reply:
for current T1 networking, we set default value for some fields, for example, subnetPrefixLength to 28.
For podCidrs/ingressCidrs etc, if they are not set on namespace settings, VAPI just leave them empty in NsxNetworkConfiguration CR, NCP will use cluster config.
I think we can use the same way for VPC settings in NSX Operator.
What do you think?

Reply:
I feel it is complex we use different strategies for different fields: some inherit Supervisor; some has default value; some (e.g. NSXProxy) are optional in VAPI but always set in CRD. Do you think possible to go one way?

Reply:
I think we may inherit Supervisor for all fields if they not set on namespace settings. Do you think this make sense?

Reply:
Sure. Maybe can go that way for now. Then all fields are optional?

Reply:
OK. Thanks Jianjun. I misunderstood your point before. I thought you mean some fields which required in NSX-T can be optional in VAPI but mandatory in this CRD, wcpsvc needs to fill these values if they are not set and inherits supervisor level settings.  But we can also set all these fields optional and inherit from Supervisor settings in NSX Operator.

Reply:
I thought that too. The problem is that if users change the Supervisor settings in VAPI, I guess not easy for wcpsvc to change all Namespaces' CRs.

Reply:
Yea, that's true.  Then we can set all fields optional and if not set, inherit from Supervisor settings in NSX Operator.